source/lac/cuda_solver_direct.cu

```diff
@@ -0,0 +1,353 @@
+                          cusolverDnHandle_t cusolver_dn_handle,
+                          const SparseMatrix<Number> &matrix,
+                          const Number *b_dev, Number *x_dev)
+    {
+      // Change the format of the matrix from sparse to dense
+      unsigned int const m = matrix.m();
+      unsigned int const n = matrix.n();
+      Assert(m == n, ExcMessage("The matrix is not square"));
+      Number *dense_matrix_dev;
+      Utilities::CUDA::malloc(dense_matrix_dev, m * n);
+
+      // Change the format of matrix to dense
+      internal::cusparsecsr2dense(cusparse_handle, matrix, dense_matrix_dev);
+
+      // Create the working space
+      int workspace_size = 0;
+      internal::cusolverDngetrf_buffer_size(cusolver_dn_handle, m, n,
+                                            dense_matrix_dev, workspace_size);
+      Assert(workspace_size > 0, ExcMessage("No workspace was allocated"));
+      Number *workspace_dev;
+      Utilities::CUDA::malloc(workspace_dev, workspace_size);
+
+      // LU factorization
+      int *pivot_dev;
+      Utilities::CUDA::malloc(pivot_dev, m);
+      int *info_dev;
+      Utilities::CUDA::malloc(info_dev, 1);
+
+      internal::cusolverDngetrf(cusolver_dn_handle, m, n, dense_matrix_dev,
+                                workspace_dev, pivot_dev, info_dev);
+
+#ifdef DEBUG
+      int info = 0;
+      cudaError_t cuda_error_code_debug =
+        cudaMemcpy(&info, info_dev, sizeof(int), cudaMemcpyDeviceToHost);
+      AssertCuda(cuda_error_code_debug);
+      Assert(info == 0, ExcMessage("There was a problem during the LU factorization"));
+#endif
+
+      // Solve Ax = b
+      cudaError_t cuda_error_code = cudaMemcpy(x_dev, b_dev, m * sizeof(Number),
+                                               cudaMemcpyDeviceToDevice);
+      AssertCuda(cuda_error_code);
+      internal::cusolverDngetrs(cusolver_dn_handle, m, dense_matrix_dev, pivot_dev,
+                                x_dev, info_dev);
+#ifdef DEBUG
+      cuda_error_code =
+        cudaMemcpy(&info, info_dev, sizeof(int), cudaMemcpyDeviceToHost);
+      AssertCuda(cuda_error_code);
+      Assert(info == 0, ExcMessage("There was a problem during the LU solve"));
+#endif
+
+      // Free the memory allocated
+      Utilities::CUDA::free(dense_matrix_dev);
+      Utilities::CUDA::free(workspace_dev);
+      Utilities::CUDA::free(pivot_dev);
+      Utilities::CUDA::free(info_dev);
+    }
+
+
+
+    template <typename Number>
+    void lu_factorization(cusolverSpHandle_t cusolver_sp_handle,
+                          const SparseMatrix<Number> &matrix,
+                          const Number *b_dev, Number *x_dev)
+    {
+      // cuSOLVER does not support LU factorization of sparse matrix on the device,
+      // so we need to move everything to the host first and then back to the host.
+      const unsigned int nnz = matrix.n_nonzero_elements();
+      const unsigned int n_rows = matrix.m();
```

Comment:
Again, to help me understand the model here: `b_dev` and `x_dev` are both pointers in device memory, and `matrix` is also (by design) on the device, right?

Reply:
Yes `b_dev` and `x_dev` are both pointers in device memory. `matrix` is our own object which has all of its data on the device.